app/api.js

```diff
@@ -91,47 +91,102 @@ export async function setPassword(id, owner_token, keychain) {
   return response.ok;
 }
 
-export function uploadFile(
+function asyncInitWebSocket(server) {
+  return new Promise(resolve => {
+    const ws = new WebSocket(server);
+    ws.onopen = () => {
+      resolve(ws);
+    };
+  });
+}
+
+async function upload(
+  ws,
+  stream,
+  streamInfo,
+  metadata,
+  verifierB64,
+  keychain,
+  onprogress,
+  canceller
+) {
+  const metadataHeader = arrayToB64(new Uint8Array(metadata));
+  const fileMeta = {
+    fileMetadata: metadataHeader,
+    authorization: `send-v1 ${verifierB64}`
+  };
+
+  function listenForResponse() {
+    return new Promise((resolve, reject) => {
+      ws.addEventListener('message', function(msg) {
+        const response = JSON.parse(msg.data);
```

Comment:
We should catch parse errors and reject the promise, otherwise the promise may never get fullfilled.